ring/batch.go

```diff
@@ -111,30 +128,56 @@ func DoBatch(ctx context.Context, op Operation, r ReadRing, keys []uint32, callb
-	// Similarly, track the number of errors, and if it exceeds maxFailures
-	// shortcut the waiting rpc.
+func (b *batchTracker) record(itemTrackers []*itemTracker, err error) {
+	// If we reach the required number of successful puts on this item, then decrement the
+	// number of pending items by one.
 	//
-	// The use of atomic increments here guarantees only a single sendSamples
-	// goroutine will write to either channel.
-	for i := range sampleTrackers {
+	// The use of atomic increments here is needed as:
+	// * rpcsPending and rpcsFailed guarantees only a single goroutine will write to either channel
+	// * succeeded, failed4xx, failed5xx and remaining guarantees that the "return decision" is made atomically
+	// avoiding race condition
+	for i := range itemTrackers {
 		if err != nil {
-			if sampleTrackers[i].failed.Inc() <= int32(sampleTrackers[i].maxFailures) {
-				continue
-			}
-			if b.rpcsFailed.Inc() == 1 {
-				b.err <- err
+			// Track the number of errors by error family, and if it exceeds maxFailures
+			// shortcut the waiting rpc.
+			errCount := itemTrackers[i].recordError(err)
+			// We should return an error if we reach the maxFailure (quorum) on a given error family OR
+			// we don't have any remaining instances to try.
+			//
+			// Ex: 2xx, 4xx, 5xx -> return 5xx
+			// Ex: 4xx, 4xx, _ -> return 4xx
+			// Ex: 5xx, _, 5xx -> return 5xx
+			//
+			// The reason for searching for quorum in 4xx and 5xx errors separately is to give a more accurate
+			// response to the initial request. So if a quorum of instances rejects the request with 4xx, then the request should be rejected
+			// even if less-than-quorum instances indicated a failure to process the request (via 5xx).
+			// The speculation is that had the unavailable instances been available,
+			// they would have rejected the request with a 4xx as well.
+			// Conversely, if a quorum of instances failed to process the request via 5xx and less-than-quorum
+			// instances rejected it with 4xx, then we do not have quorum to reject the request as a 4xx. Instead,
+			// we return the last 5xx error for debuggability.
+			if errCount > int32(itemTrackers[i].maxFailures) || itemTrackers[i].remaining.Dec() == 0 {
+				if b.rpcsFailed.Inc() == 1 {
+					b.err <- err
+				}
 			}
 		} else {
-			if sampleTrackers[i].succeeded.Inc() != int32(sampleTrackers[i].minSuccess) {
+			// If we successfully process items in minSuccess instances,
+			// then wake up the waiting rpc, so it can return early.
+			if itemTrackers[i].succeeded.Inc() >= int32(itemTrackers[i].minSuccess) {
+				if b.rpcsPending.Dec() == 0 {
+					b.done <- struct{}{}
+				}
 				continue
 			}
-			if b.rpcsPending.Dec() == 0 {
-				b.done <- struct{}{}
+
+			// If we succeeded to call this particular instance, but we don't have any remaining instances to try,
+			// and we did not succeed calling minSuccess instances, then we need to return the last error
```

Comment:
[Nit] I think this wording is simpler:
```suggestion
			// If we successfully called this particular instance, but we don't have any remaining instances to try,
			// and we failed to call minSuccess instances, then we need to return the last error
```